### PR TITLE
Add powerai-based multiarch Dockerfile

### PR DESCRIPTION
**Pull request by mattborgard**
This Dockerfile will create an empty (i.e., no deep learning frameworks) IBM PowerAI-based container with the MAX framework software installed.

Note that currently powerai 1.6.0 is gpu-based only, meaning any subsequent images based FROM this one will need access to GPU hardware (usually via nvidia-docker).

### CODE REVIEW

**Comment by frreiss** [comment]
This looks good to me. What do you think, @bdwyer2 ?

File: Dockerfile.pai
Comment (xuhdev) [inline Dockerfile.pai]: How about changing this line to `COPY . .` and remove the last line of this file?

**Comment by MLnick** [comment]
Looks fine to me.

@ajbozarth are we able to have DockerHub build an image tag off this Dockerfile in our auto build? e.g. `max-base-powerai` ?

**Comment by mattborgard** [comment]
Creating a max-base-powerai build would be great. Ideally, I'd like to make the build multiarch (x86+ppc64le). I can add the logic in your travisfile to do that, but do you know if you have ppc64le builders available?

File: Dockerfile.pai
Comment (mattborgard) [reply Dockerfile.pai]: Yep, that works. Changed.

**Comment by ajbozarth** [comment]
yeah we can set up a second auto build on a second docker file. Once this is merged I'll create a auto build for a tag called `max-base-powerai` built on the new dockerfile.

**Comment by mattborgard** [comment]
I've added a .travisfile to this PR after discussion with @ajbozarth . There's not a great way to get Docker Hub's automated build system to build multiarch images (or even just power images), so I've instead modified the travis builder to build those images.

You'll need to modify your travis builder with the following variables:

REPO: Whatever docker registry/repo we're pushing to (I'm assuming 'codait')
REPO_USER: The dockerhub user for authorizing with the above repo
REPO_PASS: And the dockerhub password

Let me know if there are any other changes here you'd prefer.

**Comment by ajbozarth** [comment]
Just on an initial glance it seems that this would push a new image to docker hub even on PR travis builds, but overwriting the master build. Am I misreading the travis file?

**Comment by mattborgard** [comment]
Good call @ajbozarth. I've added a conditional so that it doesn't push on pull requests.

**Comment by ajbozarth** [comment]
sorry I haven't had enough time to properly review this yet, I've requested other members of the MAX team to also take a look so this doesn't get blocked on me. I just want to make sure we don't let a minor edge case in the travis build mess up anything downstream since this is a large update to the yaml file.

**Comment by mattborgard** [comment]
Understandable! Let me know if I can answer any concerns. Just to be clear, this does the standard build/test for the vanilla max-base image that the previous travisfile performed, but while it pushes the new powerai based images, it does *not* push the vanilla max-base image it builds. That would still (presumably) be built by your DockerHub builder.

If y'all _want_ it pushed from Travis instead, that could of course be done.

**Comment by ptitzler** [comment]
We'll discuss today internally

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: If we move setting DOCKERFILE to `before_install` in the  os `if` statements it will allow us to remove the matrix exclude block.

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: Why do we need to set the dist to `xenial `? Is the default not ok?

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: Is using `docker manifest` necessary? The docs for it specifically say not to use it in prod environments yet. I would be more comfortable removing it if it's not a requirement for the docker image to work

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: A couple issues with this line but I'll address just one of them here then add more on other follow up lines: I don't believe there's a reason to use `"$REPO"/"$IMAGE":"$ARCH"-latest` instead of `max-base` when you can easily create the full tag inside the if block in `after_success`

Comment (ajbozarth) [reply .travis.yml]: Separately, I'm not sure `"$REPO"/"$IMAGE":"$ARCH"-latest` is the correct format to begin with. Based your naming it would resolve to `codait/max-base-powerai:ppc64le-latest` for the new build, which would try pushing to a completely new docker repo called `max-base-powerai` instead of a tag on the current `max-base` repo. Was it your intention to have a separate repo rather than just a separate tag?

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: IIUC this will not work on the first run of this file since only one of the above two pulls will succeed.

File: .travis.yml
Comment (ajbozarth) [inline .travis.yml]: For security reasons this should be `echo "$REPO_PASS" | docker login -u "$REPO_USER" --password-stdin` so the password is printed in the logs

**Review by ajbozarth — COMMENTED** [review]
I plan to create a fork and test this tomorrow (since it's EOD) but here's a handful of comments and questions for you. Also based on my comments I've made a general edit to the whole file (completely untested) to give an idea of what I'm thinking. 

```
language: python

python:
  - 3.6

os:
  - linux
  - linux-ppc64le

services:
  - docker

before_install:
  - if [[ "$TRAVIS_OS_NAME" == "linux-ppc64le" ]]; then
       DOCKERFILE=Dockerfile.pai;
    else
       DOCKERFILE=Dockerfile;
    fi

install:
  - docker build -f "$DOCKERFILE" -t max-base .

before_script:
  - docker run -it --rm -d -p 5000:5000 max-base
  - sleep 10

script:
  - docker ps | grep max-base

after_success:
  - if [[ "$TRAVIS_OS_NAME" == "linux-ppc64le" && "$TRAVIS_PULL_REQUEST" == "false" ]]; then
      echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin;
      docker tag max-base codait/max-base:powerai;
      docker push codait/max-base:powerai;
    fi
```

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: Works for me

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: I believe we used xenial on our ppc64le builders for a newer docker version. We may be able to update docker instead, or this might not be relevant any longer. I can test this.

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: So, the goal is to have a user do `docker pull max-base:powerai-latest` and get an x86 image on an x86 machine, and a power image on a power machine. This is only available through the use of the `manifest` command, as far as I know.

(Note that dockerhub supports this widely on official images at this point. Pretty much any OS image is multiarch, presumably created with the manifest command)

I understand the apprehension with turning on experimental functions. The manifest function was supposed to become a standard function, and that may have already happened in the newest version of docker. I'll check on this.

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: That is fine with me. I wrote it this way to keep the powerai-based image entirely separate from the vanilla one, but we can definitely put them in the same repo. Would you prefer something like max-base:powerai-<arch>-latest ?

The only real requirement is to make sure people know what they're getting. There won't be a non-powerai based image for ppc64le (i.e. pulling max-base:latest won't get you a ppc64le image, even on a ppc64le machine), so that needs to be clear in the documentation at the very least.

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: This is correct. The build will succeed regardless and the multiarch image will be pushed when the other arch build finishes. Also, if we manually push initial images into the repo, the first automated build will grab that. 

Now that I think about it, I believe the multiarch manifest references the component images by name/tag and not by hash. If that's the case, we might be able to manually push the multiarch manifest once, and as long as the component images keep the same name, the multiarch image will stay up to date. I can verify this.

(Note that if there's any need to tag multiarch images with a specific build or commit number instead of just `latest` we'll still need to push manifests in each build)

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: As I mentioned in https://github.com/IBM/MAX-Base/pull/13#discussion_r288382167 after thinking about this, we may not have to do this for each build after all. If we manually push a manifest, I believe it will stay updated. I will verify this.

Comment (mattborgard) [reply .travis.yml]: And that's a big no! Unfortunately there's no way around using the `manifest` command if we want a multiarch image.

File: .travis.yml
Comment (mattborgard) [reply .travis.yml]: Unfortunately I confirmed this isn't true. The manifest locks to specific shas for the component images, so a new manifest needs to be pushed with each build.

A cleaner way to do this might be to add a separate manifest creation job as a new stage that depends on the previous jobs, however, and make the `docker push` for each image build required for job success.

File: .travis.yml
Comment (ajbozarth) [reply .travis.yml]: base on our slack discussion this will not work since we need 3 builds compared to the 2 that I thought.

File: .travis.yml
Comment (ajbozarth) [reply .travis.yml]: As discussed in slack the manifest is necessary and this should stay as is

File: .travis.yml
Comment (ajbozarth) [reply .travis.yml]: `xenial` is the default as of this month: https://blog.travis-ci.com/2019-04-15-xenial-default-build-environment so no need for this line

File: .travis.yml
Comment (ajbozarth) [reply .travis.yml]: Based on our slack convo, this only needs minor changes. Give the max-base portion of this file is solely for testing, it's name in travis doesn't matter, for the power side of things this should probably be changed from `"$REPO"/"$IMAGE":"$ARCH"-latest` to `codait/max-base:"powerai-$ARCH"-latest` see my example code block in my PR comment below for my example way of replacing this

**Comment by ajbozarth** [comment]
So based on our slack convo I made some updates/suggested edits to your travis.yml and am copying them into this comment below:

```
language: python

python:
  - 3.6

os:
  - linux
  - linux-ppc64le

env:
  - DOCKERFILE=Dockerfile IMAGE=test
  - DOCKERFILE=Dockerfile.pai IMAGE=powerai

matrix:
  exclude:
  - env: DOCKERFILE=Dockerfile IMAGE=test
    os: linux-ppc64le

services:
  - docker

before_install:
  - if [[ "$TRAVIS_OS_NAME" == "linux-ppc64le" ]]; then
       ARCH=ppc64le;
    fi
  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
       ARCH=x86_64;
    fi
  - mkdir -p ~/.docker
  - echo '{"experimental":"enabled"}' | tee ~/.docker/config.json 
  - sudo service docker restart

install:
  - docker build -f "$DOCKERFILE" -t codait/max-base:"$IMAGE"-"$ARCH"-latest .

before_script:
  - docker run -it --rm -d -p 5000:5000 codait/max-base:"$IMAGE"-"$ARCH"-latest
  - sleep 10

script:
  - docker ps | grep codait/max-base:"$IMAGE"-"$ARCH"-latest

after_success:
  - if [[ "$IMAGE" != "test" && "$TRAVIS_PULL_REQUEST" == "false" ]]; then
      echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin;
      docker push codait/max-base:"$IMAGE"-"$ARCH"-latest;
    fi

jobs:
  include:
    - stage: "Docker Manifest"
      script:
      - echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin;
        docker pull codait/max-base:powerai-ppc64le-latest;
        docker pull codait/max-base:powerai-x86_64-latest;
        docker manifest create codait/max-base:powerai-latest codait/max-base:powerai-ppc64le-latest codait/max-base:powerai-x86_64-latest;
        docker manifest push codait/max-base:powerai-latest;
```

Notes based on the above code: 
- I changed how IMAGE is used and named since we want to have tags in the main max-base repo rather than a separate repo.
- The build matrix with an excludes remains since it is seemingly the best way to do the 3 builds: original test build, and the powerai build on both x86 and ppc64le
- I updated the docker login to the secure format
- I moved the manifest chunk into a new stage called "Docker Manifest" that will run after the previous 3 builds finish. Note that the powerai builds push their image in the first stage then pull both in the manifest stage

Is what I wrote look ok to you? if so feel free to update your code with it. It (and your current code) have one large flaw remaining afaik. It wont push to docker hub on PR builds or the current test build, but travis runs on and push to any branch or tag, so we need to further restrict the docker push to just master and tag releases and make the docker tag name match appropriately (ie powerai-x86_64-latest and powerai-latest vs powerai-x86_64-v1.1.2 and powerai-v1.1.2). This could most likely be done with a few if statements and setting another var in the before_install step.

**Comment by ajbozarth** [comment]
Ok so as a followup on my last comment I'v tried writing up a version with the checks for master/tag. I plan to create that fork now and test it, but here's the travis.yml I'm testing:
```
language: python

python:
  - 3.6

os:
  - linux
  - linux-ppc64le

env:
  - DOCKERFILE=Dockerfile IMAGE=test
  - DOCKERFILE=Dockerfile.pai IMAGE=powerai

matrix:
  exclude:
  - env: DOCKERFILE=Dockerfile IMAGE=test
    os: linux-ppc64le

services:
  - docker

before_install:
  - if [[ "$TRAVIS_OS_NAME" == "linux-ppc64le" ]]; then
      ARCH=ppc64le;
    fi
  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
      ARCH=x86_64;
    fi
  - if [[ "$TRAVIS_BRANCH" == "master" ]]; then
      VERSION=latest;
    else
      VERSION=$TRAVIS_BRANCH;
    fi
  - mkdir -p ~/.docker
  - echo '{"experimental":"enabled"}' | tee ~/.docker/config.json 
  - sudo service docker restart

install:
  - docker build -f "$DOCKERFILE" -t codait/max-base:"$IMAGE"-"$ARCH"-"$VERSION" .

before_script:
  - docker run -it --rm -d -p 5000:5000 codait/max-base:"$IMAGE"-"$ARCH"-"$VERSION"
  - sleep 10

script:
  - docker ps | grep codait/max-base:"$IMAGE"-"$ARCH"-"$VERSION"

after_success:
  - if [[ "$IMAGE" != "test" && "$TRAVIS_PULL_REQUEST" == "false" ]] && [[ "$TRAVIS_BRANCH" == "master" || "$TRAVIS_BRANCH" == "$TRAVIS_TAG" ]]; then
      echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin;
      docker push codait/max-base:"$IMAGE"-"$ARCH"-"$VERSION";
    fi

jobs:
  include:
    - stage: "Docker Manifest"
      if: branch = master OR tag IS present
      script:
      - if [[ "$TRAVIS_BRANCH" == "master" ]]; then
          VERSION=latest;
        else
          VERSION=$TRAVIS_BRANCH;
        fi
        echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin;
        docker pull codait/max-base:powerai-ppc64le-"$VERSION";
        docker pull codait/max-base:powerai-x86_64-"$VERSION";
        docker manifest create codait/max-base:powerai-"$VERSION" codait/max-base:powerai-ppc64le-"$VERSION" codait/max-base:powerai-x86_64-"$VERSION";
        docker manifest push codait/max-base:powerai-"$VERSION";
```

**Comment by ajbozarth** [comment]
So I forked the repo and have done some through testing of the travis/docker side of this (feel free to check out my docker images and make sure they work from the user side of things). My updated yml file can be found here: https://github.com/ajbozarth/MAX-Base/blob/master/.travis.yml

While testing I made quite a few edits to the file to get it working (from the code pasted in my above comment). I'll highlight some of them here:
- moved chunk to enable experimental mode to the stage that uses it.
- made various edits to the manifest stage such as fixing the conditional, separating out script lines for easier debugging, and making sure the stage doesn't do all the builds.

The final thing I'm still working on is that I've discovered that TRAVIS_OS_NAME=linux for both x86 and ppc64le so the current if block setting ARCH doesn't work. Based on preliminary googling we may have to completely restructure how the build matrix is written so we can differentiate between the two in the build steps

**Comment by ajbozarth** [comment]
So I decided to keep at it and I solved the os issue by using matrix includes and the docker images it built can be found here: https://cloud.docker.com/repository/docker/ajbozarth/max-base/tags

Also I found that how we were enabling manifest didn't work, but discovered and switched to a different method: force update of docker then enable via env var.

As of now the `.travis.yml` in my repo work as intended for this PR. If you can test that the images in my docker hub repo work as you intend them to then we can move forward with my updates.

**Comment by mattborgard** [comment]
This looks great, and works in my repo. You can check the images at mborgard/max-base:powerai-latest, and my basic function testing confirmed they work as expected.

As we discussed, I had to add dist: xenial to get it to build, but that seems to be an IBM Travis issue. So if you're building this elsewhere, the .travisfile currently in this PR should work.

**Review by ajbozarth — APPROVED** [review]
Sounds good, once I get the Docker Hub auth set up and then I'll merge this.

**Comment by ajbozarth** [comment]
I created the docker user `codaitbot` and filled in it's credentials in `DOCKER_USER` and `DOCKER_PASS` for both travis-ci.org and travis-ci.com, please make sure to note in each follow up PR for the models that I will need to make that change.

**Comment by ajbozarth** [comment]
No that this is merged (and since there's no open work on the repo) I will be cutting a new minor release. Any references in the models should point to the new release rather than `latest`.

I am also watching to see if having both `.org` and `.com` running will break this docker portion before cutting said release.

**Pull request merged**
